02-intermediate-ruby/lazy-with-let.md

```diff
@@ -0,0 +1,180 @@
+
+### Proof that let is lazy
+
+In this example using `before` the statement "Creating a player named Ada Lovelace" prints twice times showing that before always runs.  
+
+```ruby
+# sample_spec.rb
+
+describe "Player" do
+  before do
+  	puts "Creating a player named Ada Lovelace"
+    @player = Scrabble::Player.new "Ada Lovelace"
+  end
+
+    describe "initialize method" do
+        it "New Players initialize with a name" do
+            @player.name.must_equal "Ada Lovelace"
+        end
+
+        it "Throws an ArgumentError if created without a name" do
+          proc {
+            Scrabble::Player.new
+          }.must_raise ArgumentError
+        end
+    end
+end
+```
+Output:
+
+```bash
+$  ruby specs/sample_spec.rb
+Run options: --seed 62527
+
+# Running:
+	
+Creating a player named Ada Lovelace
+.Creating a player named Ada Lovelace
+.
+
+Finished in 0.011018s, 272.2817 runs/s, 272.2817 assertions/s.
+
+2 runs, 2 assertions, 0 failures, 0 errors, 0 skips
+```
+
+But if we use let instead:
+
+```ruby
+# player_spec.rb
+
+describe "Player" do
+  let (:player)  {
+    puts "Creating a player named Ada Lovelace"
+    Scrabble::Player.new('Ada Lovelace')
+  }
+
+
+    describe "initialize method" do
+        it "New Players initialize with a name" do
+            player.name.must_equal "Ada Lovelace"
+        end
+
+        it "Throws an ArgumentError if created without a name" do
+          proc {
+            Scrabble::Player.new
+          }.must_raise ArgumentError
+        end
+    end
+end
+```
+
+We get
+
+```bash
+ruby specs/sample_spec.rb
+Run options: --seed 49538
+
+# Running:
+
+Creating a player named Ada Lovelace
+.
+
+Finished in 0.005194s, 770.1194 runs/s, 770.1194 assertions/s.
+
+2 runs, 2 assertions, 0 failures, 0 errors, 0 skips
+```
+
+Notice the `puts` ran only once, only in the test-case where it was used.  So let can often be more efficient because it's only ran where needed.  Unlike life, in Minitest being lazy pays off!
+
+### A Longer Let Example:
+
+The below example creates 2 let statements for a player and a tilebag from Scrabble and uses them together in a batch of tests.  You can use the names defined in let like variable names.  
+
+```ruby
+describe "Player & Tilebag" do
+  let (:player)  {Scrabble::Player.new('Watson')}
+  let (:tilebag) {Scrabble::TileBag.new}
+
+  describe "draw_tiles for Player class" do
+  
+    it "fills tiles array with letters" do
+     player.draw_tiles(tilebag)
+     player.players_tiles.must_be_instance_of Array
+     player.players_tiles[0].must_be_instance_of String
+    end
+
+    it "won't fill the array beyond 7 letters" do
+      player.draw_tiles(tilebag)
+      player.players_tiles.length.must_equal 7
+    end
+
+  end
+end
+```
+
+## Summary
+
+So `let` is a useful helper method to DRY up your testing code.  With it you can define a method that is run only when used with the result cached so you can use it like a local variable in your test cases.  
```

Comment:
I'm not sure the students have the context to what it means for something to be cached. Also, I'm not sure I understand that to be true in how let is executed.